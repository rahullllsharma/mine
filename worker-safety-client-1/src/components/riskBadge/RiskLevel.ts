export enum RiskLevel {
  HIGH = "HIGH",
  MEDIUM = "MEDIUM",
  LOW = "LOW",
  UNKNOWN = "UNKNOWN",
  RECALCULATING = "RECALCULATING",
}
