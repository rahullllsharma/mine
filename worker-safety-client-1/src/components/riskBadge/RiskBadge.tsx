import { Badge, Icon } from "@urbint/silica";
import cx from "classnames";
import {
  getBackgroundColorByRiskLevel,
  getDefaultTextColorByRiskLevel,
  getRiskIcon,
} from "@/utils/risk/";
import { RiskLevel } from "./RiskLevel";

interface RiskBadgeAnimatedProps {
  label: string;
  isCritical?: boolean;
}

const RiskBadgeAnimated = ({
  label,
  isCritical = false,
}: RiskBadgeAnimatedProps): JSX.Element => {
  return (
    <div
      className="flex items-center font-bold shadow-5 h-5 rounded-full px-1.5 py-0.5 w-max uppercase whitespace-nowrap bg-risk-recalculating text-neutral-shade-75"
      role="note"
    >
      {/*<div
        className="animate-spin inline-block border-2 border-r-neutral-shade-75 w-3 h-3 mr-1 rounded-full border-neutral-shade-18"
        role="status"
      />*/}
      <span className="text-tiny m-1">{label}</span>
      {isCritical && (
        <Icon
          name={"warning"}
          className={cx(
            "pointer-events-none text-lg bg-transparent ml-1 leading-none"
          )}
        />
      )}
    </div>
  );
};

export interface RiskBadgeProps {
  risk: RiskLevel;
  label: string;
  isCritical?: boolean;
}

export default function RiskBadge({
  risk,
  label,
  isCritical = false,
}: RiskBadgeProps): JSX.Element {
  if (risk === RiskLevel.RECALCULATING) {
    return (
      <RiskBadgeAnimated label={label?.toUpperCase()} isCritical={isCritical} />
    );
  }

  return (
    <Badge
      className={cx(
        "whitespace-nowrap",
        getBackgroundColorByRiskLevel(risk),
        getDefaultTextColorByRiskLevel(risk)
      )}
      iconStart={getRiskIcon(risk)}
      label={label?.toUpperCase()}
      iconEnd={isCritical ? "warning" : undefined}
    />
  );
}
