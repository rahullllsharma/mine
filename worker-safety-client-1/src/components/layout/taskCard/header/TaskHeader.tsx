import type { IconName } from "@urbint/silica";
import type { KeyboardEventHandler } from "react";
import type { MenuItemProps } from "@/components/shared/dropdown/Dropdown";
import type { RiskLevel } from "@/components/riskBadge/RiskLevel";
import { ActionLabel, Badge, Icon } from "@urbint/silica";
import cx from "classnames";
import ButtonIcon from "@/components/shared/button/icon/ButtonIcon";
import RiskBadge from "@/components/riskBadge/RiskBadge";
import Tooltip from "@/components/shared/tooltip/Tooltip";
import Dropdown from "@/components/shared/dropdown/Dropdown";

export type TaskHeaderProps = {
  icon?: IconName;
  iconClassName?: string;
  menuIcon?: IconName;
  headerText: string;
  onClick?: () => void;
  onMenuClicked?: () => void;
  showSummaryCount?: boolean;
  totalHazards?: number;
  totalControls?: number;
  subHeaderText?: string;
  riskLevel?: RiskLevel;
  hasInfoIcon?: boolean;
  infoIconTooltipText?: string;
  hasDropdown?: boolean;
  dropdownOptions?: MenuItemProps[][];
};

export default function TaskHeader({
  icon,
  iconClassName,
  menuIcon,
  headerText,
  onClick,
  onMenuClicked,
  showSummaryCount = false,
  totalHazards = 0,
  totalControls = 0,
  subHeaderText,
  riskLevel,
  hasInfoIcon,
  infoIconTooltipText,
  hasDropdown,
  dropdownOptions = [],
}: TaskHeaderProps): JSX.Element {
  const keyPressHandler: KeyboardEventHandler<HTMLDivElement> = e => {
    // https://www.w3.org/TR/uievents-key/#keys-whitespace
    if (e.key === "Enter" || e.key === " ") {
      onClick && onClick();
    }
  };

  return (
    <div className="flex flex-1 items-stretch h-auto p-3">
      <div
        role="button"
        tabIndex={0}
        className={cx(
          "flex flex-1 flex-wrap items-stretch gap-y-3 gap-x-6 md:gap-0 md:flex-row md:flex-nowrap md:items-center mr-6 outline-none",
          { "cursor-default": !onClick }
        )}
        onClick={onClick}
        onKeyPress={keyPressHandler}
      >
        <div className="text-left text-base text-neutral-shade-100 font-bold flex flex-auto m-0 items-center gap-3 w-full md:flex-initial md:gap-4 md:w-auto md:ml-3 md:mr-4">
          {icon && (
            <Icon name={icon} className={cx("text-xl", iconClassName)} />
          )}
          <div>
            <ActionLabel className="font-semibold">{headerText}</ActionLabel>
            {subHeaderText && (
              <span className=" pl-1 text-tiny font-medium text-neutral-shade-58">
                {subHeaderText}
              </span>
            )}
          </div>
          {hasInfoIcon && (
            <Tooltip title={infoIconTooltipText ?? ""} position="right">
              <Icon name="info_circle_outline" />
            </Tooltip>
          )}
        </div>
        {riskLevel && (
          <Tooltip
            title={
              "The High / Medium / Low classification represents the relative likelihood of presence of high energy hazards while the task is being performed"
            }
            position="bottom"
            className="max-w-xl"
          >
            <RiskBadge risk={riskLevel} label={`${riskLevel}`} />
          </Tooltip>
        )}
        {showSummaryCount && (
          <div className="flex flex-1 justify-start md:justify-center ml-2">
            <Badge label={`${totalHazards}H`} className="mr-2" />
            <Badge label={`${totalControls}C`} />
          </div>
        )}
      </div>
      {/*Menu button*/}
      {!hasDropdown && onMenuClicked && menuIcon && (
        <ButtonIcon
          iconName={menuIcon}
          onClick={onMenuClicked}
          title="Menu button"
          className="flex items-center self-center px-2.5 py-2"
        />
      )}
      {hasDropdown && dropdownOptions.length > 0 && (
        <Dropdown menuItems={dropdownOptions}>
          <ButtonIcon iconName="more_horizontal" className="h-full" />
        </Dropdown>
      )}
    </div>
  );
}
