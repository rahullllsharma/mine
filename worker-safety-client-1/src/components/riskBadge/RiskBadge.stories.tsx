import type { ComponentMeta, ComponentStory } from "@storybook/react";

import RiskBadge from "./RiskBadge";
import { RiskLevel } from "./RiskLevel";

export default {
  title: "Components/Badge/RiskBadge",
  component: RiskBadge,
  argTypes: {
    riskColor: { table: { disable: true } },
  },
} as ComponentMeta<typeof RiskBadge>;

const TemplateRiskBadge: ComponentStory<typeof RiskBadge> = ({
  risk,
  label,
}) => <RiskBadge risk={risk} label={label} />;

export const Playground = TemplateRiskBadge.bind({});

Playground.args = {
  risk: RiskLevel.HIGH,
  label: "Risk Label",
};
