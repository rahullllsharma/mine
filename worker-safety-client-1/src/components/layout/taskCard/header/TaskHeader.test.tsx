import { fireEvent, render, screen } from "@testing-library/react";
import { noop } from "lodash-es";
import { RiskLevel } from "@/components/riskBadge/RiskLevel";
import TaskHeader from "./TaskHeader";

const headerTextValue = "Above Ground Welding";
const subHeaderTextValue = "sub Above Ground Welding";

describe("TaskHeader", () => {
  it("should render a label", () => {
    render(
      <TaskHeader
        icon="chevron_big_down"
        headerText="Above Ground Welding"
        onClick={noop}
      />
    );
    const labelElement = screen.getByText("Above Ground Welding");
    expect(labelElement).toBeInTheDocument();
  });

  describe("when contain sub-header", () => {
    it("should render the sub-header text", () => {
      render(
        <TaskHeader
          icon="chevron_big_down"
          headerText={headerTextValue}
          subHeaderText={subHeaderTextValue}
          onClick={noop}
        />
      );
      const labelElement = screen.getByText("sub Above Ground Welding");
      expect(labelElement).toBeInTheDocument();
    });

    it("should render the header and sub-header text", () => {
      render(
        <TaskHeader
          icon="chevron_big_down"
          headerText={headerTextValue}
          subHeaderText={subHeaderTextValue}
          onClick={noop}
        />
      );
      const headerText = screen.getByText(headerTextValue);
      const subHeaderText = screen.getByText(subHeaderTextValue);
      expect(headerText).toBeInTheDocument();
      expect(subHeaderText).toBeInTheDocument();
    });

    it("should render an icon with header and sub-header text", () => {
      const { asFragment } = render(
        <TaskHeader
          icon="chevron_big_down"
          headerText={headerTextValue}
          subHeaderText={subHeaderTextValue}
          onClick={noop}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when clickable region is clicked", () => {
    it("should call 'onClickableRegion'", () => {
      const clickableRegionHandler = jest.fn();
      render(
        <TaskHeader
          icon="chevron_big_down"
          headerText="Above Ground Welding"
          onClick={clickableRegionHandler}
        />
      );
      const button = screen.getByRole("button");
      fireEvent.click(button);
      expect(clickableRegionHandler).toHaveBeenCalled();
    });
  });

  describe("when a riskLevel is passed", () => {
    it("should render a risk badge", () => {
      const { asFragment } = render(
        <TaskHeader
          riskLevel={RiskLevel.HIGH}
          icon="chevron_big_down"
          headerText="Above Ground Welding"
          onClick={noop}
        />
      );
      const riskBadge = screen.getByRole("note");
      expect(riskBadge).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when we want to show the summary count", () => {
    it("should render two badges", () => {
      const { asFragment } = render(
        <TaskHeader
          icon="chevron_big_down"
          headerText="Above Ground Welding"
          onClick={noop}
          showSummaryCount
        />
      );
      const badges = screen.getAllByRole("note");
      expect(badges).toHaveLength(2);
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render the badge count of hazards", () => {
      render(
        <TaskHeader
          icon="chevron_big_down"
          headerText="Above Ground Welding"
          onClick={noop}
          showSummaryCount
          totalControls={0}
          totalHazards={1}
        />
      );
      const badge = screen.getByText("1H");
      expect(badge).toBeInTheDocument();
    });

    it("should render the badge count of controls", () => {
      render(
        <TaskHeader
          icon="chevron_big_down"
          headerText="Above Ground Welding"
          onClick={noop}
          showSummaryCount
          totalControls={1}
          totalHazards={0}
        />
      );
      const badge = screen.getByText("1C");
      expect(badge).toBeInTheDocument();
    });
  });

  describe("when menu button is rendered", () => {
    const menuButtonHandler = jest.fn();
    const clickableRegionHandler = jest.fn();
    const headerText = "Above Ground Welding";
    beforeEach(() => {
      render(
        <TaskHeader
          icon="chevron_big_down"
          menuIcon="edit"
          headerText={headerText}
          onClick={clickableRegionHandler}
          onMenuClicked={menuButtonHandler}
        />
      );
    });

    it("should render two buttons", () => {
      screen.getByRole("button", { name: headerText });
      screen.getByRole("button", { name: /menu button/i });
    });

    describe("when menu button is clicked", () => {
      it("should call 'onMenuClicked'", () => {
        const buttonElements = screen.getAllByRole("button");
        fireEvent.click(buttonElements[1]);
        expect(menuButtonHandler).toHaveBeenCalled();
      });
    });
  });
});
