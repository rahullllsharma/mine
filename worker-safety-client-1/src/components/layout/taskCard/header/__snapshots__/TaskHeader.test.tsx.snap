// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskHeader when a riskLevel is passed should render a risk badge 1`] = `
<DocumentFragment>
  <div
    class="flex flex-1 items-stretch h-auto p-3"
  >
    <div
      class="flex flex-1 flex-wrap items-stretch gap-y-3 gap-x-6 md:gap-0 md:flex-row md:flex-nowrap md:items-center mr-6 outline-none"
      role="button"
      tabindex="0"
    >
      <div
        class="text-left text-base text-neutral-shade-100 font-bold flex flex-auto m-0 items-center gap-3 w-full md:flex-initial md:gap-4 md:w-auto md:ml-3 md:mr-4"
      >
        <i
          aria-hidden="true"
          class="ci-chevron_big_down text-xl"
        />
        <div>
          <span
            class="text-action-label font-action-label text-neutrals-primary font-semibold"
          >
            Above Ground Welding
          </span>
        </div>
      </div>
      <span
        class="inline-flex items-center"
      >
        <span
          class="inline-flex items-center w-full"
        >
          <div
            class="flex items-center font-bold shadow-5 h-5 rounded-full p-0.5 w-max uppercase bg-brand-gray-20 whitespace-nowrap bg-risk-high text-white"
            role="note"
          >
            <span
              class="flex items-center px-1"
            >
              <i
                aria-hidden="true"
                class="ci-chevron_up"
              />
              <span
                class="text-tiny mr-1"
              >
                HIGH
              </span>
            </span>
          </div>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TaskHeader when contain sub-header should render an icon with header and sub-header text 1`] = `
<DocumentFragment>
  <div
    class="flex flex-1 items-stretch h-auto p-3"
  >
    <div
      class="flex flex-1 flex-wrap items-stretch gap-y-3 gap-x-6 md:gap-0 md:flex-row md:flex-nowrap md:items-center mr-6 outline-none"
      role="button"
      tabindex="0"
    >
      <div
        class="text-left text-base text-neutral-shade-100 font-bold flex flex-auto m-0 items-center gap-3 w-full md:flex-initial md:gap-4 md:w-auto md:ml-3 md:mr-4"
      >
        <i
          aria-hidden="true"
          class="ci-chevron_big_down text-xl"
        />
        <div>
          <span
            class="text-action-label font-action-label text-neutrals-primary font-semibold"
          >
            Above Ground Welding
          </span>
          <span
            class=" pl-1 text-tiny font-medium text-neutral-shade-58"
          >
            sub Above Ground Welding
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TaskHeader when we want to show the summary count should render two badges 1`] = `
<DocumentFragment>
  <div
    class="flex flex-1 items-stretch h-auto p-3"
  >
    <div
      class="flex flex-1 flex-wrap items-stretch gap-y-3 gap-x-6 md:gap-0 md:flex-row md:flex-nowrap md:items-center mr-6 outline-none"
      role="button"
      tabindex="0"
    >
      <div
        class="text-left text-base text-neutral-shade-100 font-bold flex flex-auto m-0 items-center gap-3 w-full md:flex-initial md:gap-4 md:w-auto md:ml-3 md:mr-4"
      >
        <i
          aria-hidden="true"
          class="ci-chevron_big_down text-xl"
        />
        <div>
          <span
            class="text-action-label font-action-label text-neutrals-primary font-semibold"
          >
            Above Ground Welding
          </span>
        </div>
      </div>
      <div
        class="flex flex-1 justify-start md:justify-center ml-2"
      >
        <div
          class="flex items-center font-bold shadow-5 h-5 rounded-full p-0.5 w-max uppercase bg-brand-gray-20 mr-2"
          role="note"
        >
          <span
            class="flex items-center px-1"
          >
            <span
              class="text-tiny mx-1"
            >
              0H
            </span>
          </span>
        </div>
        <div
          class="flex items-center font-bold shadow-5 h-5 rounded-full p-0.5 w-max uppercase bg-brand-gray-20 undefined"
          role="note"
        >
          <span
            class="flex items-center px-1"
          >
            <span
              class="text-tiny mx-1"
            >
              0C
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
